### PR TITLE
kubeadm: fix ignored `certificate-authority` in discovery file

### LINKED ISSUE #80065: kubeadm: ignored certificate-authority option in discovery file
**What happened**:
`kubeadm join` with discovery file containing `certificate-authority` option pointed to a file with CA certificate (instead of default `certificate-authority-data`) fails.

**What you expected to happen**:
`kubeadm join` with discovery file containing `certificate-authority` option pointed to a file with CA certificate successfully load the certificate from the file and use it.

**How to reproduce it (as minimally and precisely as possible)**:
For a discovery file
* extract CA certificate from `certificate-authority-data` into an external file
* remove `certificate-authority-data`
* add `certificate-authority` pointing to the external file

**Anything else we need to know?**:
nope

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.1-dirty", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"dirty", BuildDate:"2019-07-01T19:30:04Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-08T17:02:58Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
AWS
- OS (e.g: `cat /etc/os-release`):
```
NAME="Ubuntu"
VERSION="16.04.5 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.5 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```
- Kernel (e.g. `uname -a`):
`Linux kubestage1-uswest1adevc 4.4.0-1085-aws #96-Ubuntu SMP Tue Jun 11 09:08:32 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools:
`kubeadm`
- Network plugin and version (if this is a network-related bug):
N/A
- Others:

@kubernetes/sig-cluster-lifecycle-kubeadm

### PR DESCRIPTION
**Pull request by vkhromov**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Fix `kubeadm join` ignores `certificate-authority` option in a discovery file.

**Which issue(s) this PR fixes**:

Fixes #80065

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: fix ignored `certificate-authority` in discovery file
```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
this seems good to me, thanks for the fix.
(added hold for more people to see)

/lgtm
/approve
/hold
@kubernetes/sig-cluster-lifecycle-pr-reviews 
/assign @fabriziopandini

**Comment by neolit123** [comment]
@vkhromov 
nstead of `NONE`, please include the following note under `Does this PR introduce a user-facing change?:`
> kubeadm: fix ignored `certificate-authority` in discovery file

**Comment by vkhromov** [comment]
@neolit123, updated the release note

**Comment by neolit123** [comment]
thanks, let see more comments before merging.

File: cmd/kubeadm/app/discovery/file/file.go [hunk 1]
```diff
@@ -55,13 +55,23 @@ func ValidateConfigInfo(config *clientcmdapi.Config, clustername string) (*clien
 	// This is the cluster object we've got from the cluster-info kubeconfig file
 	defaultCluster := kubeconfigutil.GetClusterFromKubeConfig(config)
 
+	var caCerts []byte
```
Comment (yastij) [inline cmd/kubeadm/app/discovery/file/file.go]: maybe something `caCerts := defaultCluster.CertificateAuthorityData`

File: cmd/kubeadm/app/discovery/file/file.go [hunk 2]
```diff
@@ -55,13 +55,23 @@ func ValidateConfigInfo(config *clientcmdapi.Config, clustername string) (*clien
...
+	if defaultCluster.CertificateAuthority != "" {
+		caCerts, err = ioutil.ReadFile(defaultCluster.CertificateAuthority)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to load CA certificate referenced by kubeconfig")
+		}
+	} else {
```
Comment (yastij) [inline cmd/kubeadm/app/discovery/file/file.go]: this would get rid of the else

**Review by yastij — COMMENTED** [review]
just a nit as take or leave

**Comment by neolit123** [comment]
@vkhromov any chance you can address the above two comments?

File: cmd/kubeadm/app/discovery/file/file.go [hunk 3]
```diff
@@ -55,13 +55,23 @@ func ValidateConfigInfo(config *clientcmdapi.Config, clustername string) (*clien
 	// This is the cluster object we've got from the cluster-info kubeconfig file
 	defaultCluster := kubeconfigutil.GetClusterFromKubeConfig(config)
 
+	var caCerts []byte
+	if defaultCluster.CertificateAuthority != "" {
+		caCerts, err = ioutil.ReadFile(defaultCluster.CertificateAuthority)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to load CA certificate referenced by kubeconfig")
```
Comment (rosti) [inline cmd/kubeadm/app/discovery/file/file.go]: Maybe dumping `defaultCluster.CertificateAuthority` as part of the error message is good thing to do.

**Comment by neolit123** [comment]
https://github.com/kubernetes/kubernetes/pull/80675
is about to merge soon and will possibly cause this PR to conflict.

**Comment by neolit123** [comment]
@vkhromov ping.
since this bug fix is quite viable we still want to apply it.
let me know if you don't have the bandwidth, so that we can request a replacement PR from someone else.

File: cmd/kubeadm/app/discovery/file/file.go (hunk 1 above)
Comment (dixudx) [reply cmd/kubeadm/app/discovery/file/file.go]: Yeah, I prefer to declare like this.

**Comment by neolit123** [comment]
i will send a replacement PR for this later today.
thanks
/close

**Comment by neolit123** [comment]
new PR is here https://github.com/kubernetes/kubernetes/pull/80966

**Pull request closed without merging**